File: java/Benchmark.java
package sssp;

import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.common.AttributeKey;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
import io.opentelemetry.semconv.ResourceAttributes;

import java.io.*;
import java.nio.file.*;
import java.util.*;

/** Reads graph from file (n m then u v w lines), runs SSSP(0), prints DONE.
 * When OTEL_EXPORTER_OTLP_ENDPOINT is set, emits a trace span (e.g. for SigNoz). */
public final class Benchmark {
    public static void main(String[] args) throws IOException {
        String path = args.length > 0 ? args[0] : System.getenv("GRAPH_FILE");
        if (path == null || path.isEmpty()) {
            System.err.println("Usage: java sssp.Benchmark <graph.txt>");
            System.exit(1);
        }
        Graph g = loadGraph(path);
        String algo = System.getenv("SSSP_ALGORITHM");
        if (algo == null) algo = "duan_mao_shu_yin";
        algo = algo.trim().toLowerCase();
        int fixedIters = 0;
        String itersEnv = System.getenv("SSSP_ITERATIONS");
        if (itersEnv != null && !itersEnv.isBlank()) {
            try {
                fixedIters = Integer.parseInt(itersEnv.trim());
            } catch (NumberFormatException ignored) {}
        }
        double minSec = 0.0;
        String minSecEnv = System.getenv("SSSP_MIN_SECONDS");
        if (minSecEnv != null && !minSecEnv.isBlank()) {
            try {
                minSec = Double.parseDouble(minSecEnv.trim());
            } catch (NumberFormatException ignored) {}
        }
        double maxSec = 30.0;
        String maxSecEnv = System.getenv("SSSP_MAX_SECONDS");
        if (maxSecEnv != null && !maxSecEnv.isBlank()) {
            try {
                maxSec = Double.parseDouble(maxSecEnv.trim());
            } catch (NumberFormatException ignored) {}
        }

        OpenTelemetry otel = initOpenTelemetry();
        Tracer tracer = otel != null ? otel.getTracer("sssp-bench-java", "1.0.0") : null;
        int[] iterationsHolder = { 1 };

        if (tracer != null) {
            Span span = tracer.spanBuilder("sssp.benchmark").setAttribute("sssp.algorithm", algo).startSpan();
            try {
                runBenchmark(g, algo, fixedIters, minSec, maxSec, iterationsHolder);
            } finally {
                span.setAttribute(AttributeKey.longKey("sssp.iterations"), iterationsHolder[0]);
                span.end();
            }
        } else {
            runBenchmark(g, algo, fixedIters, minSec, maxSec, iterationsHolder);
        }

        int iterations = iterationsHolder[0];
        String resultFile = System.getenv("RESULT_FILE");
        if (resultFile != null && !resultFile.isEmpty()) {
            try {
                Files.writeString(Path.of(resultFile), "{\"iterations\":" + iterations + "}");
            } catch (IOException ignored) {}
        }
    }

    private static OpenTelemetry initOpenTelemetry() {
        String endpoint = System.getenv("OTEL_EXPORTER_OTLP_ENDPOINT");
        if (endpoint == null || endpoint.isBlank()) return null;
        try {
            OtlpGrpcSpanExporter exporter = OtlpGrpcSpanExporter.builder().build();
            String svcName = System.getenv("OTEL_SERVICE_NAME");
            if (svcName == null || svcName.isBlank()) svcName = "sssp-bench-java";
            Resource resource = Resource.getDefault().toBuilder().put(ResourceAttributes.SERVICE_NAME, svcName).build();
            SdkTracerProvider provider = SdkTracerProvider.builder()
                .addSpanProcessor(BatchSpanProcessor.builder(exporter).build())
                .setResource(resource)
                .build();
            return OpenTelemetrySdk.builder().setTracerProvider(provider).buildAndRegisterGlobal();
        } catch (Throwable t) {
            return null;
        }
    }

    private static void runBenchmark(Graph g, String algo, int fixedIters, double minSec, double maxSec, int[] iterationsHolder) {
        SsspResult r;
        if (fixedIters > 0) {
            r = "dijkstra".equals(algo) ? Dijkstra.solve(g, 0) : DuanMaoShuYinSSSP.solve(g, 0);
            for (int i = 1; i < fixedIters; i++)
                r = "dijkstra".equals(algo) ? Dijkstra.solve(g, 0) : DuanMaoShuYinSSSP.solve(g, 0);
            iterationsHolder[0] = fixedIters;
        } else if (minSec > 0) {
            long startNanos = System.nanoTime();
            long minNanos = (long) (minSec * 1_000_000_000);
            long maxNanos = (long) (maxSec * 1_000_000_000);
            int iters = 0;
            do {
                r = "dijkstra".equals(algo) ? Dijkstra.solve(g, 0) : DuanMaoShuYinSSSP.solve(g, 0);
                iters++;
            } while (System.nanoTime() - startNanos < minNanos && System.nanoTime() - startNanos < maxNanos);
            iterationsHolder[0] = iters;
        } else {
            r = "dijkstra".equals(algo) ? Dijkstra.solve(g, 0) : DuanMaoShuYinSSSP.solve(g, 0);
        }
    }

    static Graph loadGraph(String path) throws IOException {
        try (Scanner sc = new Scanner(new File(path))) {
            int n = sc.nextInt(), m = sc.nextInt();
            Graph g = new Graph(n);
            for (int i = 0; i < m; i++)
                g.addEdge(sc.nextInt(), sc.nextInt(), sc.nextDouble());
            g.compact();
            return g;
        }
    }
}

File: rust/src/bin/benchmark.rs
// Load graph from file, run SSSP(0), print DONE. Path = argv[1] or GRAPH_FILE env.
use std::env;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;
use std::time::Instant;

use sssp::{dijkstra, duan_mao_shu_yin, Graph};

fn load_graph(path: &Path) -> std::io::Result<Graph> {
    let f = File::open(path)?;
    let mut lines = BufReader::new(f).lines();
    let first = lines.next().ok_or_else(|| std::io::Error::from(std::io::ErrorKind::InvalidData))??;
    let mut parts = first.split_whitespace();
    let n: usize = parts.next().unwrap().parse().unwrap();
    let m: usize = parts.next().unwrap().parse().unwrap();
    let mut g = Graph::new(n);
    for _ in 0..m {
        let line = lines.next().ok_or_else(|| std::io::Error::from(std::io::ErrorKind::InvalidData))??;
        let mut p = line.split_whitespace();
        let u: usize = p.next().unwrap().parse().unwrap();
        let v: usize = p.next().unwrap().parse().unwrap();
        let w: f64 = p.next().unwrap().parse().unwrap();
        g.add_edge(u, v, w);
    }
    Ok(g)
}

fn main() {
    let path = env::args()
        .nth(1)
        .or_else(|| env::var("GRAPH_FILE").ok())
        .unwrap_or_else(|| {
            eprintln!("Usage: benchmark <graph.txt>");
            std::process::exit(1);
        });
    let path = Path::new(&path);
    let g = load_graph(path).unwrap_or_else(|e| {
        eprintln!("{}", e);
        std::process::exit(1);
    });
    let algo = env::var("SSSP_ALGORITHM").unwrap_or_else(|_| "duan_mao_shu_yin".into());
    let min_sec: f64 = env::var("SSSP_MIN_SECONDS")
        .ok()
        .and_then(|s| s.trim().parse().ok())
        .unwrap_or(0.0);
    let max_sec: f64 = env::var("SSSP_MAX_SECONDS")
        .ok()
        .and_then(|s| s.trim().parse().ok())
        .unwrap_or(30.0);
    if min_sec > 0.0 {
        let start = Instant::now();
        let min_dur = std::time::Duration::from_secs_f64(min_sec);
        let max_dur = std::time::Duration::from_secs_f64(max_sec);
        let mut iters = 0u64;
        let mut r = if algo.trim().eq_ignore_ascii_case("dijkstra") {
            dijkstra(&g, 0)
        } else {
            duan_mao_shu_yin(&g, 0)
        };
        while start.elapsed() < min_dur && start.elapsed() < max_dur {
            r = if algo.trim().eq_ignore_ascii_case("dijkstra") {
                dijkstra(&g, 0)
            } else {
                duan_mao_shu_yin(&g, 0)
            };
            iters += 1;
        }
        let reachable = r.distance.iter().filter(|d| d.is_finite()).count();
        println!("DONE {} {} {}", r.vertex_count(), reachable, iters);
    } else {
        let r = if algo.trim().eq_ignore_ascii_case("dijkstra") {
            dijkstra(&g, 0)
        } else {
            duan_mao_shu_yin(&g, 0)
        };
        let reachable = r.distance.iter().filter(|d| d.is_finite()).count();
        println!("DONE {} {} 1", r.vertex_count(), reachable);
    }
}
